### PR TITLE
Add IP and Referrer restriction support to API Key authentication

### PR DESCRIPTION
**Pull request by ChamodDamitha**
API keys can be restricted to be used only by

1. Specific IP address
  - By specifying the addresses exactly or giving the address range using CIDR notation)

2. Specific Http referrers (Web sites)
 - By specifying URLs exactly or using wildcards

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+
+  const onIpTextUpdate = (e) => {
+    updateNewIp(e.target.value.trim());
+    if (e.target.value.trim() == "") {
+      setInvalidIP(false);
+    }
+  };
+
+  const addIpItem = () => {
+    if (newIP != null && newIP != "") {
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:101]: Shal we use `!==` here
example.
```javascript
12 != '12' // false
```

This evaluates to `false`

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        ].join("")
+      );
+
+      const ipv6RegExp = new RegExp(
+        [
+          "^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]",
+          "{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:",
+          "[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4})",
+          "{1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|",
+          "([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]",
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:116]: We already have [`@hapi`](https://github.com/wso2/carbon-apimgt/blob/master/features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/package.json#L22) for validation requirements , Shall we use its [ip validation function](https://hapi.dev/module/joi/api/?v=17.1.1#stringipoptions) ?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+                  margin="dense"
+                  variant="outlined"
+                  placeholder="Enter IP address"
+                  fullWidth
+                />
+              </Grid>
+              <Grid item md={2} xs={2}>
+                <span>
+                  <Fab
+                    style={{ marginLeft: "20px", marginRight: "20px" }}
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:215]: Please use [material ui theme spacing](https://material-ui.com/customization/spacing/)

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+                  margin="dense"
+                  variant="outlined"
+                  placeholder="Enter Referer"
+                  fullWidth
+                />
+              </Grid>
+              <Grid item md={2} xs={2}>
+                <span>
+                  <Fab
+                    style={{ marginLeft: "20px", marginRight: "20px" }}
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:285]: Same here, Please use https://material-ui.com/customization/spacing/

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx [hunk 5]
```diff
@@ -139,60 +178,141 @@ class ApiKeyManager extends React.Component {
...
-                                id='Shared.AppsAndKeys.ViewKeys.consumer.close.btn'
-                                defaultMessage='Close'
-                            />
+                        {showToken && <ViewToken token={apikey} />}
+                      </DialogContentText>
+                    </DialogContent>
+                    <DialogActions>
+                      {!showToken && (
+                        <Button
+                          onClick={() => this.generateKeys(selectedApp, keyType)}
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx:242]: This is an anti-pattern, We have created a lambda function and called a class instance method without using any event data ?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx [hunk 6]
```diff
@@ -139,60 +178,141 @@ class ApiKeyManager extends React.Component {
...
+                        </Typography>
+                        <List>
+                          <ListItem>
+                            <Typography variant="body1" component="p">
+                              Specify one IPv4 or IPv6 or a subnet using CIDR notation
+                            </Typography>
+                          </ListItem>
+                          <ListItem>
+                            <Typography variant="body1" component="p">
+                              Examples: <b>192.168.1.2</b>, <b>152.12.0.0/13</b>,{" "}
```
Comment (tmkasun) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx:276]: Please add i18n support in this text and all the other places as well , For more info about [wso2 i18n](https://github.com/wso2/carbon-apimgt/wiki/How-internationalization-(i18n)-works-in-API-Manager-React-Apps)

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+import Box from "@material-ui/core/Box";
+import Radio from "@material-ui/core/Radio";
+import RadioGroup from "@material-ui/core/RadioGroup";
+import FormControl from "@material-ui/core/FormControl";
+import Button from "@material-ui/core/Button";
+
+// Styles for Grid and Paper elements
+const styles = (theme) => ({
+  FormControl: {
+    "margin-bottom": "8px",
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:43]: Shall we use theme.spacing here? e.g: theme.spacing(1)

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    width: "100%",
+    padding: "0px 10px",
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:45]: use theme.spacing(3)

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+  };
+
+  const deleteRefererItem = (refererItem) => {
+    refererList.splice(refererList.indexOf(refererItem), 1);
+    updateRefererList(refererList);
+  };
+
+  const onIpTextUpdate = (e) => {
+    updateNewIp(e.target.value.trim());
+    if (e.target.value.trim() == "") {
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:95]: shall we use === ?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+    setInvalidIP(false);
+    setInvalidReferer(false);
+  };
+
+  return (
+    <React.Fragment>
+      <Box
+        border={1}
+        borderRadius={5}
+        style={{ margin: "10px" , padding: "10px"}}
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:158]: Don't use inline styling. Let's define styles at the top.

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+        borderColor="grey.500"
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:159]: Don't use custom colors. The theme has pre-defined colors. Use them

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx [hunk 12]
```diff
@@ -139,60 +178,141 @@ class ApiKeyManager extends React.Component {
...
+                      </CardContent>
+                    </Card>
+                  </Grid>
+                )}
+                {restrictSchema == "referer" && (
+                  <Grid item md={5} xs={12}>
+                    <Card>
+                      <CardContent>
+                        <Typography variant="h5" component="h2">
+                          Examples of URLs allowed to restrict websites
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx:290]: Add i18n support

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx [hunk 13]
```diff
@@ -139,60 +178,141 @@ class ApiKeyManager extends React.Component {
...
+                          </ListItem>
+                          <ListItem>
+                            <Typography variant="body1" component="p">
+                              Any URL in a single subdomain, using a wildcard asterisk (*):{" "}
+                              <b>sub.example.com/*</b>
+                            </Typography>
+                          </ListItem>
+                          <ListItem>
+                            <Typography variant="body1" component="p">
+                              Any subdomain or path URLs in a single domain, using wildcard
```
Comment (kavishkafernando) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx:306]: Add i18n support

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 1 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 2 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 3 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 4 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx (hunk 5 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx (hunk 6 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 7 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 8 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 9 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 10 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 11 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx (hunk 12 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx]: fixed

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx (hunk 13 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/AppsAndKeys/ApiKeyManager.jsx]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java [hunk 14]
```diff
@@ -358,6 +361,71 @@ public AuthenticationResponse authenticate(MessageContext synCtx) {
         }
     }
 
+    private void validateAPIKeyRestrictions(JWTClaimsSet payload, MessageContext synCtx) throws APISecurityException {
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) synCtx).
+                getAxis2MessageContext();
+
+        String permittedIPList = (String) payload.getClaim(APIConstants.JwtTokenConstants.PERMITTED_IP);
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:368]: null check claim not available at none state

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java [hunk 15]
```diff
@@ -358,6 +361,71 @@ public AuthenticationResponse authenticate(MessageContext synCtx) {
...
+
+        if (StringUtils.isNotEmpty(permittedIPList)) {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:370]: != null and not empty

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java [hunk 16]
```diff
@@ -339,7 +339,17 @@ public Response applicationsApplicationIdApiKeysKeyTypeGeneratePost(
...
                         validityPeriod = -1;
                     }
-                    String apiKey = apiConsumer.generateApiKey(application, userName, (long) validityPeriod);
+
+                    String restrictedIP = null;
+                    String restrictedReferer = null;
+
+                    if (body.getAdditionalProperties() != null) {
+                        Map additionalProperties = (HashMap) body.getAdditionalProperties();
+                        restrictedIP = (String) additionalProperties.get(APIConstants.JwtTokenConstants.PERMITTED_IP);
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java:348]: null check and string

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/ApiKeyGenerator.java [hunk 17]
```diff
@@ -86,6 +86,9 @@ private static String buildBody(JwtTokenInfoDTO jwtTokenInfoDTO) {
         jwtClaimsSetBuilder.claim("application", jwtTokenInfoDTO.getApplication());
         jwtClaimsSetBuilder.claim("keytype", jwtTokenInfoDTO.getKeyType());
 
+        jwtClaimsSetBuilder.claim(APIConstants.JwtTokenConstants.PERMITTED_IP, jwtTokenInfoDTO.getPermittedIP());
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/ApiKeyGenerator.java:89]: null check and set

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java (hunk 15 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:370]: StringUtils.isNotEmpty() method already checks for null. So, why do we need to check again?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java (hunk 14 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:368]: null can be type cast to String and it still will be null. So, do we need to check for null here?

Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java (hunk 16 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/ApiKeyGenerator.java (hunk 17 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/token/ApiKeyGenerator.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java [hunk 18]
```diff
@@ -358,6 +361,78 @@ public AuthenticationResponse authenticate(MessageContext synCtx) {
...
+
+                    if (StringUtils.isNotEmpty(clientIP)) {
+                        log.debug("Invocations to API: " + apiContext + ":" + apiVersion +
+                                " is not permitted for client with IP: " + clientIP);
+                    }
+                }
+                throw new APISecurityException(APISecurityConstants.API_AUTH_FORBIDDEN,
+                        "Access forbidden for the invocations");
+            }
+
```
Comment (Meruja) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:396]: Shall we remove this extra line?

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java [hunk 19]
```diff
@@ -358,6 +361,78 @@ public AuthenticationResponse authenticate(MessageContext synCtx) {
...
+            // Validate http referer against the permitted referrers
+            TreeMap<String, String> transportHeaderMap = (TreeMap<String, String>)
+                    axis2MessageContext.getProperty
+                            (org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+            if (transportHeaderMap != null) {
+                String referer = transportHeaderMap.get("Referer");
+                if (StringUtils.isNotEmpty(referer)) {
+                    for (String restrictedReferer : permittedRefererList.split(",")) {
+                        String restrictedRefererRegExp = restrictedReferer.trim()
+                                .replace("*", "[^ ]*");
```
Comment (ruwiniwj) [inline components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:414]: can't we move this to the previous line? does it exceed 120 line length?

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+            <Grid
+              container
+              direction="row"
+              spacing={0}
+              justify="left"
+              alignItems="left"
+              md={10}
+              xs={10}
+            >
+              {ipList.length > 0 && (
```
Comment (ruwiniwj) [inline features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:230]: need to have a space before open curly and after close curly. Fix elsewhere as well.

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java (hunk 18 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:396]: Fixed in https://github.com/wso2/carbon-apimgt/pull/9250

File: components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java (hunk 19 above)
Comment (ChamodDamitha) [reply components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/apikey/ApiKeyAuthenticator.java:414]: Fixed in https://github.com/wso2/carbon-apimgt/pull/9250

File: features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx (hunk 20 above)
Comment (ChamodDamitha) [reply features/apimgt/org.wso2.carbon.apimgt.store.feature/src/main/resources/devportal/source/src/app/components/Shared/ApiKeyRestriction.jsx:230]: Fixed in https://github.com/wso2/carbon-apimgt/pull/9250

**Pull request merged**